File: LeetCode/77Combinations/Combinations.cpp
class Solution {
public:
    void backtrack(int n, int k, vector<vector<int>> &ans, int start, vector<int> curr) {
        if (curr.size() == k) {
            ans.push_back(curr);
            return;
        }
        for (int i = start; i <= n; i++) {
            curr.push_back(i);
            backtrack(n, k, ans, i+1, curr);
            curr.pop_back();
        }
    }
    vector<vector<int>> combine(int n, int k) {
        // Backtracking -> Time:O(n^k), Space:O(k)
        vector<vector<int>> ans;
        vector<int> curr;
        backtrack(n, k, ans, 1, curr);
        return ans;
    }
};

File: LeetCode/77Combinations/Combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        # Backtracking -> Time:O(n^k), Space:O(k)
        ans = []
        def backtrack(start_num, curr):
            if len(curr) == k:
                ans.append(curr)
                return
            for i in range(start_num, n+1):
                backtrack(i+1, curr + [i])
        backtrack(1, [])
        return ans
